In [32]:
import speech_recognition as sr
import threading

- Import speech_recognition for processing audio and converting it into text
- Import threading for multithread processing

In [33]:
recognizer = sr.Recognizer()
recognizer.energy_threshold = 300

- Instantiate speech_recognition and store them as recognizer.
- Set energy threshold to 300. So if the values are higher than the energy threshold = 300 then are considered as speech but if the values are lower then they are considered as silent. (300 is the recomended value on SpeechRecognition)

In [34]:
curThread = []
end = False

- curThread global variable for threads.
- end global boolean variable that tell real time audio recognition when to stop.

In [35]:
def recognizeAudio(audio_text):
    global end
    global curThread
    if not end:
        try:
            text = recognizer.recognize_google(audio_data=audio_text, show_all=True)
            for prediction in text['alternative']:
                print(" " + prediction["transcript"] + " (" + str(prediction["confidence"]*100) + "%)")
                if prediction["transcript"] == 'exit':
                    end = True
        except:
            print("Unable to recognize voice")

This method takes recorded audio as parameter. Then convert it into text with SpeechRecognizer's recognize_google. I picked recognized_google because it doesn't require anymore installation or api keys. This is the first one I tried and the result came out great so I decided to stick with it.
- If sucess the it print the converted audio text. 
- If sucess and the converted audio text is exit then set global variable end to true.
- If fail then print "Unable to recognize voice"

In [36]:
def theadController(audio_text):
    global end
    global curThread
    if not end:
        if len(curThread) == 0:
            curThread.append(threading.Thread(target=recognizeAudio, args=(audio_text,)))
            curThread[-1].start()
        else:
            for thread in curThread:
                if thread.is_alive():
                    curThread.append(threading.Thread(target=recognizeAudio, args=(audio_text,)))
                    curThread[-1].start()
                    break
                else:
                    thread = threading.Thread(target=recognizeAudio, args=(audio_text,))
                    thread.start()
                    break

This method manages thread, with recorded audio as parameter.
- If the global variables curThread is enpty then start a new thread that execute the method recognizeAudio with recorded audio as parameter.
- If a thread is not running in curThread then start a new thread that execute the method recognizeAudio with recorded audio as parameter.
- If a thread is running in curThread then append a new thread that execute the method recognizeAudio with recorded audio as parameter.

In [37]:
processStart = False
while not end:
    recognizer = sr.Recognizer()
    recognizer.energy_threshold = 300
    with sr.Microphone(sample_rate=16000) as source:
        print('listening')
        audio_text = recognizer.listen(source)
        print('stop listening')
        if processStart:
            if process.is_alive():
                process.join()
        process = threading.Thread(target=theadController, args=(audio_text,))
        process.start()
        processStart = True

listening
stop listening
listening
 hello (97.219545%)
stop listening
listening
 how are you (97.219545%)
stop listening
listening
 artificial intelligence (97.219545%)
stop listening
listening
 computer (97.219545%)
stop listening
listening
 listened for more second to calibrate the energy threshold (95.161003%)
Unable to recognize voice
stop listening
listening
Unable to recognize voice
stop listening
listening
 my name is Adam (89.654261%)
Unable to recognize voice
stop listening
listening
 exit (97.219545%)
stop listening


Contuniously listen for audio with SpeechRecognition.Microphone. When it is done recording then execute a thread that called theadController with recorded audio as parameter.

I set sample_rate for SpeechRecognition.Microphone as 16000Htz because this is the lowest recomended rate on SpeechRecognition. I want the lowest rate for faster processing, and after verious testing the result cames out very good.

The program will not start any more thread if it transcribed the word exit.

This is the result of one of my run: <br/>
listening <br/>
stop listening <br/>
listening <br/>
 hello (97.219545%) <br/>
stop listening <br/>
listening <br/>
 how are you (97.219545%) <br/>
stop listening <br/>
listening <br/>
 artificial intelligence (97.219545%) <br/>
stop listening <br/>
listening <br/>
 computer (97.219545%) <br/>
stop listening <br/>
listening <br/>
 listened for more second to calibrate the energy threshold (95.161003%) <br/>
Unable to recognize voice <br/>
stop listening <br/>
listening <br/>
Unable to recognize voice <br/>
stop listening <br/>
listening <br/>
 my name is Adam (89.654261%) <br/>
Unable to recognize voice <br/>
stop listening <br/>
listening <br/>
 exit (97.219545%) <br/>
stop listening <br/>